gateway/src/service.rs

```diff
@@ -0,0 +1,522 @@
+
+        // TODO I don't understand the API for `headers`: it gives an
+        // impl. of `Header` which can only be encoded in something
+        // that `Extend<HeaderValue>` but `HeaderMap` only impls
+        // `Extend<(HeaderName, HeaderValue)>` (as one would expect),
+        // therefore why the ugly hack below.
+        {
+            use axum::headers::Header;
+            let auth_header = Authorization::basic(&control_key, "");
+            let auth_header_name = Authorization::<Basic>::name();
+            let mut auth = vec![];
+            auth_header.encode(&mut auth);
+            let headers = req.headers_mut();
+            headers.remove(auth_header_name);
+            headers.append(auth_header_name, auth.pop().unwrap());
+        }
+
+        if !route.starts_with("/") {
+            route = format!("/{route}");
+        }
+
+        route = format!("/projects/{project_name}{route}");
+
+        *req.uri_mut() = route.parse().unwrap();
+
+        let target_url = format!("http://{target_ip}:8001");
+
+        debug!("routing control: {target_url}");
+
+        let resp = hyper_reverse_proxy::call("127.0.0.1".parse().unwrap(), &target_url, req)
+            .await
+            .map_err(|_| Error::kind(ErrorKind::ProjectUnavailable))?;
+
+        Ok(resp)
+    }
+
+    async fn iter_projects(&self) -> Result<impl Iterator<Item = Work>, Error> {
+        let iter = query("SELECT * FROM projects")
+            .fetch_all(&self.db)
+            .await?
+            .into_iter()
+            .map(|row| Work {
+                project_name: row.get("project_name"),
+                work: row.get::<SqlxJson<Project>, _>("project_state").0,
+                account_name: row.get("account_name"),
+            });
+        Ok(iter)
+    }
+
+    pub async fn find_project(&self, project_name: &ProjectName) -> Result<Project, Error> {
+        query("SELECT project_state FROM projects WHERE project_name=?1")
+            .bind(project_name)
+            .fetch_optional(&self.db)
+            .await?
+            .map(|r| {
+                r.try_get::<SqlxJson<Project>, _>("project_state")
+                    .unwrap()
+                    .0
+            })
+            .ok_or_else(|| Error::kind(ErrorKind::ProjectNotFound))
+    }
+
+    async fn update_project(
+        &self,
+        project_name: &ProjectName,
+        project: &Project,
+    ) -> Result<(), Error> {
+        query("UPDATE projects SET project_state = ?1 WHERE project_name = ?2")
+            .bind(&SqlxJson(project))
+            .bind(project_name)
+            .execute(&self.db)
+            .await?;
+        Ok(())
+    }
+
+    pub async fn key_from_account_name(&self, account_name: &AccountName) -> Result<Key, Error> {
+        let key = query("SELECT key FROM accounts WHERE account_name = ?1")
+            .bind(account_name)
+            .fetch_optional(&self.db)
+            .await?
+            .map(|row| row.try_get("key").unwrap())
```

Comment:
Can use `get` instead of `try_get` to avoid repeating column name and unwrapping. Perhaps overkill but may be worth implementing `FromRow` on `Key` and then using `query_as`.
```suggestion
            .map(|row| row.get(0))
```